pkg/runs/runsDownload.go

```diff
@@ -29,19 +30,47 @@ func DownloadArtifacts(
 
 	var err error = nil
 	var runs []galasaapi.Run
-	var artifactPaths []string
 
-	runs, err = GetRunsFromRestApi(runName, 0, 0, timeService, apiServerUrl)
+	if (err == nil) && (runName != "") {
+		err = ValidateRunName(runName)
+	}
 	if err == nil {
-		for _, run := range runs {
-			runId := run.GetRunId()
-			artifactPaths, err = GetArtifactPathsFromRestApi(runId, apiServerUrl)
-			if err == nil {
-				for _, artifactPath := range artifactPaths {
-					var artifactData io.Reader
-					artifactData, err = GetFileFromRestApi(runId, strings.TrimPrefix(artifactPath, "/"), apiServerUrl)
+		runs, err = GetRunsFromRestApi(runName, 0, 0, timeService, apiServerUrl)
+		if err == nil {
+			if len(runs) > 1 {
+				// get list of runs that are reRuns - get list of runs that are reRuns of each other
+				// create a map of lists of reRuns - key is queued time, value is the run
+				reRunsByQueuedTime := createMapOfReRuns(runs)
+
+				err = downloadReRunArtfifacts(reRunsByQueuedTime, forceDownload, fileSystem, apiServerUrl, console, timeService)
+
+			} else if len(runs) == 1 {
+				err = downloadArtifactsToDirectory(apiServerUrl, runName, runs[0], fileSystem, forceDownload, console)
+			} else {
+				log.Printf("No artifacts to download for run: '%s'", runName)
+				err = galasaErrors.NewGalasaError(galasaErrors.GALASA_INFO_NO_ARTIFACTS_TO_DOWNLOAD, runName)
+			}
+		}
+	}
+	return err
+}
+
+func downloadReRunArtfifacts(
+	reRunsByQueuedTime map[string][]galasaapi.Run,
+	forceDownload bool,
+	fileSystem utils.FileSystem,
+	apiServerUrl string,
+	console utils.Console,
+	timeService utils.TimeService) error {
+	var err error = nil
+	for _, reRunsList := range reRunsByQueuedTime {
+		if err == nil {
+			for reRunIndex, reRun := range reRunsList {
+				if err == nil {
+					directoryName := nameReRunArtifactDownloadDirectory(reRun, reRunIndex, timeService)
+					err = downloadArtifactsToDirectory(apiServerUrl, directoryName, reRun, fileSystem, forceDownload, console)
 					if err == nil {
-						err = WriteArtifactToFileSystem(fileSystem, runName, artifactPath, artifactData, forceDownload)
+						err = console.WriteString("Created folder: '" + directoryName + "'\n")
```

Comment:
Is there an assert statement in the unit test to check this ? Couldn't see it...

Reply:
I have on line 665 in runsDownload_test. but I will add for the other tests